010-anonymous-login.md

```diff
@@ -0,0 +1,114 @@
+
+### Prior art
+
+[prior-art]: #prior-art
+
+Luckily we have a lot of resources and examples around the internet that we can take as an example. I don't think I need to name them here, since a simple Google search will satisfy everyone's curiosity.
+
+<!--
+
+Discuss prior art, both the good and the bad, in relation to this proposal. A
+few examples of what this can include are:
+
+- Does this functionality exist in other software and what experience has their
+  community had?
+- For other teams: What lessons can we learn from what other communities have
+  done here?
+- Papers: Are there any published papers or great posts that discuss this? If
+  you have some relevant papers to refer to, this can serve as a more detailed
+  theoretical background.
+
+This section is intended to encourage you as an author to think about the
+lessons from other software, provide readers of your RFC with a fuller picture.
+If there is no prior art, that is fine - your ideas are interesting to us
+whether they are brand new or if it is an adaptation from other software.
+
+Write your answer below.
+-->
+
+### Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+**What happens if the user wants to convert his anonymous account to his already existing user account?**
+
+To be fair, I have no idea how to handle this properly. We could convert his data to his old account and also transfer all his stuff to the old user id. But this would break consistency, if there are any references in the database to the anonymous user id.
+
+**If we add one more way of authentication - it will get messy!**
+
+As mentioned in our meeting before, we might want to decouple the user accounts from email authentication. Meaning, that an account can have multiple ways of authentication. This way we could also disable users from abusing the endpoint of creating an account with an E-Mail address if this is not wanted from the owner of an Appwrite project. 
+
+For example:
+
+I create a new project and in the settings I can choose what authentication methods I want to use. Let's say I want to use E-Mail and Google OAuth2. A user creates an account via the the `POST /v1/account/email` endpoint. This will create an user with that E-Mail and Password as one possible way of logging into this account. Now the user also wants to connect his Google Account via the `GET /v1/account/oauth2/{provider}` endpoint. This endpoint will recognize a already logged in user due to cookies and will add that authentication method to the account. Now all of a sudden the owner needs to add GitHub as an OAuth2 provider to attract more developers. This would also allow multiple E-Mail addresses for a single account.
+
+Also, if we disable E-Mail authentication - this way we can ensure users not creating such accounts over the REST API.
+
```

Comment:
Some of this concerns are addressed in my previous comment, at least the ones about multiple auth methods (email + sms).

OAuth login in Appwrite already checks for existing user before creating a new one to avoid multiple users with the same email.

I agree we should allow to disable authentication method from the console. We can adjust the OAuth providers screen to have checkboxes for email/pass and anon login methods at the top. Those settings can be part of the project object.

Reply:
What do you think about decoupling users from the E-Mail authentication and treat every way fo authentication like OAuth2 providers right now?

It would end up to something like this:

<img width="1438" alt="Bildschirmfoto 2021-02-15 um 11 53 54" src="https://user-images.githubusercontent.com/1759475/107937708-9587c000-6f84-11eb-8746-1277617e06f8.png">
<img width="1436" alt="Bildschirmfoto 2021-02-15 um 11 50 56" src="https://user-images.githubusercontent.com/1759475/107937726-9a4c7400-6f84-11eb-88d0-a3da64c4e20b.png">

Reply:
Not saying it needs to be done for this RFC, just something to think about for future implementations.

Reply:
First screenshot, I think we should just add an attribute to the session and couple between the session and the login type = session type, the is also a plan in the backlog to attach access tokens to the session object.

Second screenshot, let do as part of this feature. I would call the tab `Settings` at least for until we'll have more stuff to split.

Should we turn email/pass and anon login by default for new projects?

Reply:
Yes, turning them both on by default should be fine 👍🏻 

I'll look into implementing it and adapt the RFC 🙂

Reply:
@eldadfux some thoughts:

With this we shouldn't use the E-Mail address as an identifier for linking OAuth2 logins to a user account. Especially since this value may not be unique to this user and is not suitable for use as a primary key. For this we should use the `sub` value provided by the OAuth2 providers 👍🏻 

Also, maybe the E-Mail address used for the Appwrite user is not the same for the Google Account. This would involve adding a whole new collection, maybe called `authentications` that will be used for authenticating the user and is also refered to in the user's `tokens` for example. This involves heavy refactoring and needs to be discussed and maybe its own RFC.

Reply:
If we would abstract every authentication we would also have a very good starting point for any new authentication method added in the future.